### PR TITLE
Fix numba AdvancedIncSubtensor1 with broadcasted values

### LINKED ISSUE #93: No broadcasting support in numba AdvancedIncSubtensor1
### Description

```python
import pytensor
import pytensor.tensor as pt
import numpy as np

x = pt.dvector("z")
idx = pt.ivector("idx")

out = pt.inc_subtensor(x[idx], 1)  # The one needs to be broadcasted

func = pytensor.function([x, idx], out, mode="NUMBA")
func(np.zeros(3), np.array([2], dtype=np.int32))
# -> TypingError in numba
```

This works with the C backend.

### PR DESCRIPTION
**Pull request by ricardoV94**
## Related Issue

- [x] Closes #93 

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [x] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

File: pytensor/link/numba/dispatch/basic.py [hunk 1]
```diff
@@ -597,36 +597,71 @@ def numba_funcify_IncSubtensor(op, node, **kwargs):
...
 
     if set_instead_of_inc:
-        advancedincsubtensor1_inplace = global_numba_func(
-            advancedincsubtensor1_inplace_set
-        )
+        if explicit_broadcast:
+
+            @numba_njit(boundscheck=True)
+            def advancedincsubtensor1_inplace(x, val, idxs):
+                val = val[0]
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/basic.py:614]: I couldn't find a way to distinguish between the case where we need to index and that we do not in a single njitted function, if anybody knows how to let me know.

I tried checking ndim, but if it's a float it doesn't have that property. I tried `isinstance(val, np.ndarray)` but numba doesn't like that either.

File: pytensor/link/numba/dispatch/basic.py [hunk 2]
```diff
@@ -597,36 +597,58 @@ def numba_funcify_IncSubtensor(op, node, **kwargs):
...
+    # TODO: Add explicit expand_dims in make_node so we don't need to worry about this here
+    broadcast = vals.type.ndim < x.type.ndim or vals.type.broadcastable[0]
 
     if set_instead_of_inc:
-        advancedincsubtensor1_inplace = global_numba_func(
-            advancedincsubtensor1_inplace_set
-        )
+        if broadcast:
+
+            @numba_njit(boundscheck=True)
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/basic.py:611]: @aseyboldt  is there an advantage to having the function on the global scope, and do we want it even if we need 4 versions like it seems we do?

Comment (aseyboldt) [reply pytensor/link/numba/dispatch/basic.py:611]: Probably not. The original reason I put them there was that I thought that would make it easier for numba to cache the functions. I'm not really sure if that is even true though, I still don't fully understand numbas caching system though.

**Comment by aseyboldt** [comment]
I think we might need a workaround for this issue: https://github.com/numba/numba/issues/9573
If the value on the right hand side of the update is a rank zero array (ie not a scalar) that gets broadcasted, I *think* the current code will break. We should at least test for that I think. Maybe by manually calling funcify and because of the recent no_cpython_wrapper addition, a manual wrapper function around that?

**Comment by ricardoV94** [comment]
> I think we might need a workaround for this issue: [numba/numba#9573](https://github.com/numba/numba/issues/9573) If the value on the right hand side of the update is a rank zero array (ie not a scalar) that gets broadcasted, I _think_ the current code will break. We should at least test for that I think. Maybe by manually calling funcify and because of the recent no_cpython_wrapper addition, a manual wrapper function around that?

The last test case I added in this PR is a tensor scalar on the rhs:

```python
(
            pt.as_tensor(np.arange(3 * 4 * 5).reshape((3, 4, 5))),
            pt.as_tensor(rng.poisson(size=())),  # <- Scalar y
            ([2, 0],),
        ),
```

**Comment by ricardoV94** [comment]
Isn't this line the workaround already?
```python
                if not isinstance(val, np.number) and val.ndim == x.ndim:
                    val = val[0]
```

**Comment by ricardoV94** [comment]
Okay finally got it, the bug is when the lhs is a vector, and the rhs a 0d array. It's fine if the lhs is a higher rank array, which was what I was testing before

**Comment by ricardoV94** [comment]
@aseyboldt I added a patch for the bug you found and a test that would fail without the patch. Let me know what you think

File: pytensor/link/numba/dispatch/basic.py [hunk 3]
```diff
@@ -604,36 +604,70 @@ def numba_funcify_IncSubtensor(op, node, **kwargs):
...
+            def advancedincsubtensor1_inplace(x, val, idxs):
+                if val.ndim == x.ndim:
+                    core_val = val[0]
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/basic.py:621]: I have to reassign to a new named variable, otherwise typing fails. Is that expected from Numba?

Comment (aseyboldt) [reply pytensor/link/numba/dispatch/basic.py:621]: Sometimes numba seems to require this when the type changes, but I don't know the exact rules either.

**Review by aseyboldt — APPROVED** [review]
Looks good

File: pytensor/link/numba/dispatch/basic.py (hunk 3 above)
Comment (ricardoV94) [reply pytensor/link/numba/dispatch/basic.py:621]: Opened an issue, perhaps it's documented behavior, seemed strange though: https://github.com/numba/numba/issues/9587

Comment (ricardoV94) [inline pytensor/link/numba/dispatch/basic.py:621]: This is always safe because x must be at least 1D (otherwise `make_node` raises)

**Pull request merged**